tools/osbuilder/rootfs-builder/rootfs.sh

```diff
@@ -652,26 +577,156 @@ detect_host_distro()
+
+
+	# Setup Dockerfile
+	generate_dockerfile "${distro_config_dir}"
+
+	# Build container image for building the rootfs/agent:
+	"$container_engine" build  \
+		--build-arg http_proxy="${http_proxy}" \
+		--build-arg https_proxy="${https_proxy}" \
+		-t "${image_name}" "${distro_config_dir}"
+
+
+	# fake mapping if KERNEL_MODULES_DIR is unset
+	kernel_mod_dir=${KERNEL_MODULES_DIR:-${ROOTFS_DIR}}
+
+	docker_run_args=""
+	docker_run_args+=" --rm"
+	docker_run_args+=" --runtime ${DOCKER_RUNTIME}"
+
+	if [ -z "${AGENT_SOURCE_BIN}" ] ; then
+		docker_run_args+=" --env RUST_AGENT_PKG=${RUST_AGENT_PKG}"
+		docker_run_args+=" --env RUST_AGENT=${RUST_AGENT}"
+	else
+		docker_run_args+=" --env AGENT_SOURCE_BIN=${AGENT_SOURCE_BIN}"
+		docker_run_args+=" -v ${AGENT_SOURCE_BIN}:${AGENT_SOURCE_BIN}"
+		docker_run_args+=" -v ${GOPATH_LOCAL}:${GOPATH_LOCAL} --env GOPATH=${GOPATH_LOCAL}"
+	fi
+
+	docker_run_args+=" $(docker_extra_args $distro)"
+
+	# Relabel volumes so SELinux allows access (see docker-run(1))
+	if command -v selinuxenabled > /dev/null && selinuxenabled ; then
+		SRC_VOL=("${GOPATH_LOCAL}")
+		for volume_dir in "${script_dir}" "${ROOTFS_DIR}" "${kernel_mod_dir}" "${SRC_VOL[@]}"; do
+			chcon -Rt svirt_sandbox_file_t "$volume_dir"
+		done
+	fi
+
+	before_starting_container
+	trap after_stopping_container EXIT
+
+	#Make sure we use a compatible runtime to build rootfs
+	# In case Clear Containers Runtime is installed we dont want to hit issue:
+	#https://github.com/clearcontainers/runtime/issues/828
+	"$container_engine" run  \
+		--env https_proxy="${https_proxy}" \
+		--env http_proxy="${http_proxy}" \
+		--env AGENT_VERSION="${AGENT_VERSION}" \
+		--env ROOTFS_DIR="/rootfs" \
+		--env AGENT_BIN="${AGENT_BIN}" \
+		--env AGENT_INIT="${AGENT_INIT}" \
+		--env KERNEL_MODULES_DIR="${KERNEL_MODULES_DIR}" \
+		--env EXTRA_PKGS="${EXTRA_PKGS}" \
+		--env OSBUILDER_VERSION="${OSBUILDER_VERSION}" \
+		--env INSIDE_CONTAINER=1 \
+		--env SECCOMP="${SECCOMP}" \
+		--env DEBUG="${DEBUG}" \
+		--env HOME="/root" \
+		-v "${repo_dir}":"/kata-containers" \
+		-v "${ROOTFS_DIR}":"/rootfs" \
+		-v "${kernel_mod_dir}":"${kernel_mod_dir}" \
+		$docker_run_args \
+		${image_name} \
+		bash /kata-containers/tools/osbuilder/rootfs-builder/rootfs.sh "${distro}"
+}
+
 main()
 {
-	parse_arguments $*
+	parse_arguments "$@"
 	check_env_variables
 
+	local container_engine
+	if [ -n "${USE_DOCKER}" ]; then
+		container_engine="docker"
+	elif [ -n "${USE_PODMAN}" ]; then
+		container_engine="podman"
+	fi
+
+	if [ -n "$container_engine" ]; then
+		build_with_container "${container_engine}" "${distro}"
+
+		# Alpine can't build the rust agent. To work around this,
+		# we "skip" this part when building in the container. Let's 
+		# build in that situation, otherwise our work here is done!
+		if [ "$distro" == "alpine" ]; then
+			info "Doing direct build of the rust agent for Alpine"
+			init="${ROOTFS_DIR}/sbin/init"
+			setup_rootfs
+		fi
+		exit $?
```

Comment:
Here `$?` maybe not safe. The return value of `build_with_container` will be overwritten by  `if [ "$distro" == "alpine" ]`. 
Maybe changed to something like this, depends on what error you want to return:

```bash
build_with_container "${container_engine}" "${distro}"
ret=$?

if [ "$distro" == "alpine" ]; then
	info "Doing direct build of the rust agent for Alpine"
	init="${ROOTFS_DIR}/sbin/init"
	setup_rootfs
	ret=$?
fi

exit $ret
```